autoprotocol/protocol.py

```diff
@@ -5080,51 +5081,62 @@ def provision(self, resource_id, dests, volumes):
         dests = WellGroup(dests)
         if not isinstance(resource_id, str):
             raise TypeError("Resource ID must be a string.")
-        if not isinstance(volumes, list):
-            volumes = [Unit(volumes)] * len(dests)
+        if not isinstance(amounts, list):
+            amounts = [Unit(amounts)] * len(dests)
         else:
-            if len(volumes) != len(dests):
+            if len(amounts) != len(dests):
                 raise RuntimeError(
                     "To provision a resource into multiple "
-                    "destinations with multiple volumes, the  "
-                    "list of volumes must correspond with the "
+                    "destinations with multiple volumes/masses, the  "
+                    "list of volumes/masses must correspond with the "
                     "destinations in length and in order."
                 )
-            volumes = [Unit(v) for v in volumes]
-        for v in volumes:
-            if not isinstance(v, (str, Unit)):
-                raise TypeError("Volume must be a string or Unit.")
+            amounts = [Unit(v) for v in amounts]
+        unique_measure_modes = set()
+        for amount in amounts:
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError("Volume/Mass must be a string or Unit.")
+            if amount.unit in VOLUMETRIC_UNITS:
+                unique_measure_modes.add("volume")
+            elif amount.unit in MASS_UNITS:
+                unique_measure_modes.add("mass")
```

Comment:
If I understand the intention behind this change, the intention here is to ensure that we generate units which are part of the Autoprotocol-spec.

If that's the intention, I think it'll make more sense to check this during the final serialization portion, i.e. in https://github.com/autoprotocol/autoprotocol-python/blob/master/autoprotocol/protocol.py#L5810 or perhaps upstream in the `as_dict` call.

That way, it'll be more applied to all Unit references, not just provision. 

If we were to do that, something else we need to think about is we could potentially automatically convert it to the smallest/largest supported unit. For example, someone specifying "1000:nanoliter" is technically valid when converted to "1:microliter", and they're equivalent.